test/unit/Cardano/DBLayer/MVarSpec.hs

```diff
@@ -1,14 +1,139 @@
+import Test.QuickCheck.Monadic
+    ( monadicIO )
 
+import qualified Data.Set as Set
+import qualified Data.Text as T
 
 spec :: Spec
-spec = return ()
+spec = do
+    describe "DB works as expected" $ do
+        it "readCheckpoints works properly"
+            (checkCoverage dbReadCheckpointsProp)
+        it "replacement of values works properly"
+            (checkCoverage dbReplaceValsProp)
+        it "multiple sequential putCheckpoints work properly"
+            (checkCoverage dbMultiplePutsSeqProp)
+        it "multiple parallel putCheckpoints work properly"
+            (checkCoverage dbMultiplePutsParProp)
+
+
+{-------------------------------------------------------------------------------
+                                    Properties
+-------------------------------------------------------------------------------}
+
+
+dbReadCheckpointsProp
+    :: (PrimaryKey WalletId, Int)
+    -> Property
+dbReadCheckpointsProp (key, val)  = monadicIO $ liftIO $ do
+    db <- newDBLayer
+
+    putCheckpoints db key (toWalletState val)
+    resFromDb <- readCheckpoints db key
+
+    resFromDb `shouldBe` (Just $ toWalletState val)
+
+
+dbReplaceValsProp
+    :: (PrimaryKey WalletId, Int, Int)
+    -> Property
+dbReplaceValsProp (key, val1, val2)  = monadicIO $ liftIO $ do
+    db <- newDBLayer
+
+    putCheckpoints db key (toWalletState val1)
+    putCheckpoints db key (toWalletState val2)
+    resFromDb <- readCheckpoints db key
+
+    resFromDb `shouldBe` (Just $ toWalletState val2)
+
+dbMultiplePutsSeqProp
+    :: KeyValPairs
+    -> Property
+dbMultiplePutsSeqProp (KeyValPairs keyValPairs) = monadicIO $ liftIO $ do
+    db <- newDBLayer
+
+    mapM_ (\(key, val) -> putCheckpoints db key (toWalletState val)) keyValPairs
+    resFromDb <- Set.fromList <$> readWallets db
+
+    resFromDb `shouldBe` (Set.fromList (map fst keyValPairs))
+
+dbMultiplePutsParProp
+    :: KeyValPairs
+    -> Property
+dbMultiplePutsParProp (KeyValPairs keyValPairs) = monadicIO $ liftIO $ do
+    db <- newDBLayer
+
+    mapConcurrently_ (\(key, val) -> putCheckpoints db key (toWalletState val)) keyValPairs
+    resFromDb <- Set.fromList <$> readWallets db
+
+    resFromDb `shouldBe` (Set.fromList (map fst keyValPairs))
+
+
+{-------------------------------------------------------------------------------
+                      Tests machinery, Arbitrary instances
+-------------------------------------------------------------------------------}
+
+
+newtype KeyValPairs = KeyValPairs [(PrimaryKey WalletId, Int)]
+    deriving (Show, Eq)
+
+instance Arbitrary KeyValPairs where
+    -- No shrinking
+    arbitrary = do
+        pairs <- choose (10, 50) >>= flip vectorOf arbitrary
+        KeyValPairs <$> pure pairs
+
+toWalletState
+    :: (IsOurs s, Semigroup s, NFData s, Show s) => s
+    -> NonEmpty (Wallet s)
+toWalletState val = initWallet val :| []
+
+instance Eq (Wallet Int) where
+    Wallet _ _ _ s1 == Wallet _ _ _ s2
+        = s1 == s2
```

Comment:
This is dodgy. We can probably just force the wallet state `s` to have an `Eq` and `Ord` instance, and derive this constraint for any `Wallet s` in the primitive itself.

Reply:
derived `Eq` instance in `Wallet.hs`